### PR TITLE
gh-101072: support default and kw default in PyEval_EvalCodeEx  for 3.11+

### PR DESCRIPTION
**Pull request by MatthieuDartiailh**
Fixes the regression observed in PyEval_EvalCodeEx on 3.11+ and add a test to avoid future regressions.

This should be backported to 3.11 I believe.

* Issue: gh-101072

### CODE REVIEW

**Review by kumaraditya303 — DISMISSED** [review]
Please format the C code as per https://peps.python.org/pep-0007/

**Comment by MatthieuDartiailh** [comment]
I have made the requested changes; please review again

**Review by ambv — CHANGES_REQUESTED** [review]
👍🏻 Fix works.
👍🏻 New tests fail without the fix applied.
👎🏻 There's a refleak in the new test.

Before (`main` at 7a253103d4c):
```
0:00:00 load avg: 3.77 [1/1] test_capi
beginning 9 repetitions
123456789
.........
test_capi passed in 49.2 sec

== Tests result: SUCCESS ==
```

After (this PR rebased on top of 7a253103d4c):
```
0:00:00 load avg: 3.72 [1/1] test_capi
beginning 9 repetitions
123456789
.........
test_capi leaked [1, 1, 1, 1] references, sum=4
test_capi failed (reference leak) in 49.6 sec

== Tests result: FAILURE ==
```

File: Python/ceval.c [hunk 1]
```diff
@@ -1761,9 +1761,6 @@ PyEval_EvalCodeEx(PyObject *_co, PyObject *globals, PyObject *locals,
         }
         allargs = newargs;
     }
-    for (int i = 0; i < kwcount; i++) {
-        PyTuple_SET_ITEM(kwnames, i, Py_NewRef(kws[2*i]));
-    }
```
Comment (ambv) [inline Python/ceval.c:1766]: This code is duplicated in the block right above, leading to double new refs.

**Comment by MatthieuDartiailh** [comment]
Thanks for the edits @ambv !

**Review by ambv — APPROVED** [review]

File: Modules/_testcapimodule.c [hunk 2]
```diff
@@ -3028,6 +3028,144 @@ eval_get_func_desc(PyObject *self, PyObject *func)
...
+        goto exit;
+    }
+
+
+    result = PyEval_EvalCodeEx(
+        code,
+        globals,
+        locals,
+        c_args,
+        c_args_len,
```
Comment (kumaraditya303) [inline Modules/_testcapimodule.c:3152]: @ambv Looks like there are compiler warnings. See L-3152

Comment (ambv) [reply Modules/_testcapimodule.c:3152]: Good catch, looks like `PyEval_EvalCodeEx` lists those arguments simply as `int`s. I'll convert this code to do the same.

Comment (ambv) [reply Modules/_testcapimodule.c:3152]: Nikita reported this as GH-101656.

**Pull request merged**